crates/oxc_transformer/src/es2016/exponentiation_operator.rs

```diff
@@ -0,0 +1,322 @@
+pub struct ExponentiationOperator<'a> {
+    _ctx: Ctx<'a>,
+    var_declarations: std::vec::Vec<Vec<'a, VariableDeclarator<'a>>>,
+}
+
+#[derive(Debug)]
+struct Exploded<'a> {
+    reference: AssignmentTarget<'a>,
+    uid: Expression<'a>,
+}
+
+impl<'a> ExponentiationOperator<'a> {
+    pub fn new(ctx: Ctx<'a>) -> Self {
+        Self { _ctx: ctx, var_declarations: vec![] }
+    }
+
+    fn clone_identifier_reference(
+        ident: &IdentifierReference<'a>,
+        ctx: &mut TraverseCtx<'a>,
+    ) -> IdentifierReference<'a> {
+        let reference = ctx.symbols().get_reference(ident.reference_id.get().unwrap());
+        let symbol_id = reference.symbol_id();
+        let flag = reference.flag();
+        ctx.create_reference_id(ident.span, ident.name.clone(), symbol_id, *flag)
+    }
+
+    fn clone_expression(expr: &Expression<'a>, ctx: &mut TraverseCtx<'a>) -> Expression<'a> {
+        match expr {
+            Expression::Identifier(ident) => ctx
+                .ast
+                .expression_from_identifier_reference(Self::clone_identifier_reference(ident, ctx)),
+            _ => expr.clone_in(ctx.ast.allocator),
+        }
+    }
+
+    pub fn transform_statements(
+        &mut self,
+        _statements: &mut Vec<'a, Statement<'a>>,
+        ctx: &mut TraverseCtx<'a>,
+    ) {
+        self.var_declarations.push(ctx.ast.vec());
+    }
+
+    pub fn transform_statements_on_exit(
+        &mut self,
+        statements: &mut Vec<'a, Statement<'a>>,
+        ctx: &mut TraverseCtx<'a>,
+    ) {
+        if let Some(declarations) = self.var_declarations.pop() {
+            if declarations.is_empty() {
+                return;
+            }
+            let variable = ctx.ast.alloc_variable_declaration(
+                SPAN,
+                VariableDeclarationKind::Var,
+                declarations,
+                false,
+            );
+            statements.insert(0, Statement::VariableDeclaration(variable));
+        }
+    }
+
+    pub fn transform_expression(&mut self, expr: &mut Expression<'a>, ctx: &mut TraverseCtx<'a>) {
+        // left ** right
+        if let Expression::BinaryExpression(binary_expr) = expr {
+            if binary_expr.operator == BinaryOperator::Exponential {
+                let left = ctx.ast.move_expression(&mut binary_expr.left);
+                let right = ctx.ast.move_expression(&mut binary_expr.right);
+                *expr = Self::math_pow(left, right, ctx);
+            }
+        }
+
+        // left **= right
+        if let Expression::AssignmentExpression(assign_expr) = expr {
+            if assign_expr.operator == AssignmentOperator::Exponential {
+                let mut nodes = ctx.ast.vec();
+                let Some(Exploded { reference, uid }) =
+                    self.explode(&mut assign_expr.left, &mut nodes, ctx)
+                else {
+                    return;
+                };
```

Comment:
Where left hand side is a member expression, I think we can avoid cloning it. We could instead get an owned copy of `left` here with `ast.move_expression(&assign_expr.left)`. Then insert that in new position, with no cloning.

```js
// In
obj.prop **= exp;

// Out
// NB: only 1 x `obj` - no cloning needed
var _obj;
(_obj = obj), (_obj.prop = Math.pow(_obj.prop, exp));
```

In above example, there's only 1 `obj` in the output, so no need for cloning. I think this also holds for all `MemberExpression` cases.

In the simpler case where left hand side is just an identifier, we *do* need to clone it:

```js
// In
value **= exp;

// Out
// NB: 2 x `value`s - cloning needed
var _value;
(_value = value), (value = Math.pow(_value, exp))
```

but cloning an identifier is much simpler than cloning an expression.